### PR TITLE
BOLT 2: send channel_type in open/accept

### PR DESCRIPTION
**Pull request by rustyrussell**
This is extracted from channel_upgrade (#868), but used for opening
negotiation as suggested by @roasbeef on the last spec meeting.

It's a trivial change, fully backwards compatible, but now each channel
has a channel_type, which defines its behavior, rather than an ad-hoc
set of "sticky" feature bits.  It also means both peers can *support* a
feature without endorsing it.

**Note**: `channel_type` reuses feature bits so it's trivial to know how to extend in future (e.g. option_simplified_update, option_taproot, etc).  It's not an arbitrary set of feature bits, there really are only 4 values currently defined!

(I'm tempted to make another commit which changes the rest of the text to refer explicitly to channel_type now).

### CODE REVIEW

**Review by t-bast — COMMENTED** [review]
Thanks for putting this together, this will be really useful.
It looks good to me, I'll implement in eclair this week.

File: 02-peer-protocol.md [hunk 1]
```diff
@@ -305,6 +325,9 @@ funding transaction and both versions of the commitment transaction.
     1. type: 0 (`upfront_shutdown_script`)
     2. data:
         * [`...*byte`:`shutdown_scriptpubkey`]
+    1. type: 1 (`channel_type`)
+    2. data:
+        * [`channel_type`:`type`]
```
Comment (t-bast) [inline 02-peer-protocol.md:330]: NB: it's a bit wasteful here to reuse `channel_type`, because the length will be specified twice (once because it's a TLV, and then again inside the TLV with the `u16`:`length`).

**Comment by TheBlueMatt** [comment]
Concept ACK.

File: 02-peer-protocol.md [hunk 2]
```diff
@@ -197,6 +206,13 @@ know this node will accept `funding_satoshis` greater than or equal to 2^24.
 Since it's broadcast in the `node_announcement` message other nodes can use it to identify peers 
 willing to accept large channel even before exchanging the `init` message with them. 
 
+Channel features are explicitly enumerated as `channel_type`
+bitfields, using odd features bits.  The currently defined types are:
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply 02-peer-protocol.md:210]: Because we already have one mechanism to coordinate, let's not create another!  For example, I already have variants which add option_simplified_update to all of these....

File: 02-peer-protocol.md (hunk 1 above)
Comment (t-bast) [reply 02-peer-protocol.md:330]: But I guess it prevents us from reusing existing codecs as is, so probably not worth it for a couple bytes...

Comment (rustyrussell) [reply 02-peer-protocol.md:330]: True, but it's pretty neat inside the code, as we're dealing with channel_type everywhere ("do these channel types match?").  This is especially true since we autogenerate our code from the spec, as does lnprototest.

File: 02-peer-protocol.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply 02-peer-protocol.md:210]: > Because we already have one mechanism to coordinate

If we don't want to have to assign a new set of bits, we can just use the current bit position number for the actual feature bits themselves. Since it's just an integer, we don't waste any space by not assigning non-contiguous values. To simplify things a bit the optional bit can be used, so static key is 13, anchor zero htlc is 23, etc, etc. 

The other issue with using the raw feature bit vector here is the size of the message. This likely isn't the last TLV we'll add here, and given it's an embedded feature vector it can swell to several KBs which means the message may run up against the current max message size.

Using a set integer based off the feature bit position addresses all these issues and doesn't attempt to send a redundant set of bytes to communicate things we already know. If we've already sent our feature bits in the node ann, and in innit upon connection, why do we need to send them again here? I already know what you support!  

> what do I do if I receive a byte array contains bits 7, 13, and 21? Or if I receive bits 13, 21 and 37? The first one is harmless, the second one should probably be rejected

This example shows how sending the feature bits again doesn't actually make the negotiation _explicit_ and just adds more feature bit compatibility issues at a second site (first in the node ann, and now here again). Having the acceptor "pick" amongst an advertised set just adds more moving parts, and also makes weird downgrade cases (that we've grappled w/ in the past) exist at another site (what does "most preferred" even mean?). 

If I send type `13` (static key), then it's 100% unambiguous what type of channel I want to open, no extra mapping (set of feature bits to a type) or implicit negotiation needed. It's dead simple. 

> For example, I already have variants which add option_simplified_update to all of these....

This is an argument for just using a single type imo. Lets say CL implement that new commitment update type, when newer more distinct commitment types (witness asymmetric commitments, eltoo, all the taproot variants, hybrid DLC etc) arise, that protocol will likely need to be revamped to support the new HTLC/commitment variants. Rather than needing to interpret that "modification" bit, a _new_ commitment type should be added instead to enumerate all the possible supported update+commitment variants. This makes things _explicit_ and gets rid of the weird (and error prone) implicit negotiation we have atm. This enumeration doesn't add any in-protocol overhead as a `bigsize` integer can be used. 

[We've already implemented it](https://github.com/lightningnetwork/lnd/pull/5373) and just using that type directly is a lot simpler than needing to map the set of feature bits (and determine incompatibilities or ambiguities, etc) to a type that ultimately needs to exist anyway for any implementation.

[the remark this answers is not in the document]
Comment (t-bast) [reply 02-peer-protocol.md:210]: I've thought about this more, and the issue with a `channel_type` enum as done by lnd (and by myself at first) is the explosion of enum values when you combine orthogonal features (**the `channel_type` isn't only about the commitment format, it's really the channel's feature set**).

Right now we have:

- "basic" channels
- static remotekey channels
- anchor output channels
- anchor output with 0-fee htlc channels

If we add two features that are orthogonal to the commitment format, for example `option_simplified_update` and `option_zero_reserve` (these are just examples, imagine two channel features that are completely orthogonal to the commitment format), the enum becomes:

- "basic" channels
- static remotekey channels
- anchor output channels
- anchor output with 0-fee htlc channels
- "basic" channels with simplified update
- static remotekey channels with simplified update
- anchor output channels with simplified update
- anchor output with 0-fee htlc channels with simplified update
- "basic" channels with zero reserve
- static remotekey channels with zero reserve
- anchor output channels with zero reserve
- anchor output with 0-fee htlc channels with zero reserve
- "basic" channels with zero reserve and simplified update
- static remotekey channels with zero reserve and simplified update
- anchor output channels with zero reserve and simplified update
- anchor output with 0-fee htlc channels with zero reserve and simplified update

That simply doesn't scale...and is correctly handled by rusty's proposal IMHO. In the channel DB we'd just store the feature bytes array corresponding to channel features (features that affect the channel parameters/format). As discussed during yesterday's spec meeting, each `channel_type` is uniquely identified by the value of its `features` field so it's not ambiguous.

I don't think the size of the message is a real issue. Even if we have 150 feature bits (which will take decades), it means each `channel_type` is ~20 bytes. If your implementation supports 50 distinct `channel_type`s (which is probably unreasonable from a security point of view) that's only 1kB. It can be an issue if you want to use features that aren't in the spec though, and that need to use a high feature bit. In that case it can be a bit higher, but if you use a feature bit in the 150-300 range it's still not an issue imo.

Each implementation would probably internally define the list of `channel_type`s they want to support, ordered by preference (e.g. I'd like to only support static_remotekey and anchor outputs, with or without option_simplified_update, and I'd prefer using anchor outputs and option_simplified_update) and somehow let the node operator configure what she wants to allow/disallow, but how to present that to the node operator is orthogonal to how we internally negotiate it in the channel open protocol and store it in our channel database.

File: 02-peer-protocol.md [hunk 3]
```diff
@@ -130,6 +130,15 @@ the funding transaction and both versions of the commitment transaction.
     1. type: 0 (`upfront_shutdown_script`)
     2. data:
         * [`...*byte`:`shutdown_scriptpubkey`]
+    1. type: 1 (`channel_types`)
+    2. data:
+        * [`...*channel_type`:`types`]
+
+1. subtype: `channel_type`
+2. data:
+    * [`u16`:`len`]
```
[the remark this answers is not in the document]
Comment (TheBlueMatt) [reply 02-peer-protocol.md:139]: Because if its a u16 you can serialize an existing Features object (or at least we can), but if its a bigsize we need a separate serializer for features in this context.

**Review by t-bast — COMMENTED** [review]
Concept ACK, I've prototyped this and I'm now convinced it's more future-proof to use a features field instead of defining a new enum.

I'm wondering whether we should use even or odd feature bits, it feels to me that **even** feature bits make more sense here (these features **must** be applied to the channel) but I don't feel strongly about that.

Note that there is a small implementation subtlety with regards to commit feerate. If when using anchor outputs you use a *lower* commit feerate than you would use for standard channels, and you want to offer `channel_types=[standard, anchor_outputs]`, you must use the commit feerate of a standard channel in your `open_channel` message. If your peer chooses to use anchor outputs, then you will be able to send an `update_fee` to lower the feerate. But you shouldn't offer the anchor output feerate before knowing what channel type your peer will select, otherwise you may end up with a standard channel with a feerate that's too low to confirm on its own, and if your peer plays dumb and ignores your future `update_fee` you may waste time getting your funds back.

File: 02-peer-protocol.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply 02-peer-protocol.md:210]: > is the explosion of enum values when you combine orthogonal features

Yeah this is the tradeoff in terms of the "enum space": everything needs to be unrolled. However there isn't any added overhead on the wire given that you can pack just about every combination you can think of into a `bigsize` integer. 

> That simply doesn't scale..

I guess it depends where you think the scale issues arise. It's simply taking up space already allocated (bit wise) in the integer specified. In the end, if you send all the individual feature bits, you'll end up needing to wrangle with the same "parameter space" IMO, as you need to map those down to an actual channel type (concrete implementation).

One alternative would be to split things into a base channel type, and a modifier. In your example, the modifiers would be the zero reserve and the simplified update channel type. 

I think in the end though, all implementations simply won't support every combination of the feature bit expansion (we shouldn't underestimate the complexity here). As an example, lnd doesn't support making a "basic" channel any longer, though we need to _continue_ to signal the optional bit, as otherwise with the way feature bits are set up right now, things break down on the connection level. As more and more channel types are added (taproot variants, eltoo, DLC, etc, etc) IMO it'll become more and more difficult for an implementation to support all of them. Having an "enum" for the channel type simply makes that expansion _explicit_. 

> I don't think the size of the message is a real issue.
> If your implementation supports 50 distinct channel_types (which is probably unreasonable from a security point of view) that's only 1kB

IMO the issue here is that as the TLV expands in this message, it may become impossible to honestly advertise the set of parameters you want, as you risk running into the max message limit. Why send kilobytes when a 4 bytes will do? It's unnecessarily inefficient and made worse by the fact that we don't use a sparse encoding. 

The other issue I see here as mentioned above is that we're just sending the exact same feature bit information that's already present and has been exchanged potentially twice at this point (in the init and node_ann message). On top of this, by letting the _responder_ ultimately select the channel type, the initiator is unable to actually pin down the channel type they want to send, other than sending a single feature bit (or set of bits), which reduces down to simply sending the channel type. 

On top of that, if I want to open an explicit channel type (I only want the "most secure" option), why would I ever send an array of options and allow the receiver to _choose_ amongst them? It's unclear how to advise implementations to create a safe generalized "channel type satisficement" engine (accepting the "best" set of bits). I don't think we should easily dismiss the complexity added here w.r.t preference and selection. 

In short I see future potential for a series of inadvertent footguns introduced by adding a loose "negotiation layer". An implementation needs to wrangle with parsing all the different combinations, and choosing which one is "best" for it (to advertise and accept) vs just sending what they want (based on the feature bits already advertised at the othe layers) and the other side rejecting or accepting.

**Comment by Roasbeef** [comment]
> Note that there is a small implementation subtlety with regards to commit feerate.

I see this as just one of the many future edge cases that need to be wrangled with by introducing this loose negotiation layer. If the type I want to used is just sent out right, I can just set the params in the `open_chan` message accordingly and don't need to worry about how those fields need to be modified or generalized in order to prepare for what the receiver might accept.

**Comment by Roasbeef** [comment]
> Thinking about this out a bit further, I just don't see how it's possible to craft a generalized open_channel message that is able to encompass the set of channel type possibilities in the future.

AFAICT, in order for this to actually work (this loose negotiation layer vs explicit channel types), things need to happen in two phases (kinda like the client+server hello in TLS before the actual encrypted session creation): first both sides agree on the channel type to be used, then the initial `open_channel` message is sent containing the proper paramaters to kick off channel funding. 

If channel funding is actually explicit (funder sends over type they want to used, recveiver takes it or leaves it), this additional negotiation layer isn't needed.

File: 02-peer-protocol.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (t-bast) [reply 02-peer-protocol.md:210]: > One alternative would be to split things into a base channel type, and a modifier. In your example, the modifiers would be the zero reserve and the simplified update channel type.

I agree, but these modifiers are exactly the same thing as feature bits, aren't they? So why introduce a different mechanism?

> IMO the issue here is that as the TLV expands in this message, it may become impossible to honestly advertise the set of parameters you want, as you risk running into the max message limit. Why send kilobytes when a 4 bytes will do? It's unnecessarily inefficient and made worse by the fact that we don't use a sparse encoding.

In my example, 1kB is really the most extreme case, I'm pretty sure this will never use more than ~100 bytes in the next couple of years. IMHO the number of bytes used in the messages isn't a strong enough argument to weigh in favor of one proposal or the other.

> I think in the end though, all implementations simply won't support every combination of the feature bit expansion

I completely agree, but I can really see a future where there a few distinct orthogonal feature bits that an implementation may support, and where this mechanism is useful (an implementation could very well support two distinct update protocols, two distinct commitment formats and a few minor "modifiers").

> On top of that, if I want to open an explicit channel type (I only want the "most secure" option), why would I ever send an array of options 

You don't have to. If you want only one `channel_type`, you should send just that one in your message. But if other implementations want to let the fundee choose among a range of options, we should give them the freedom to do that, shouldn't we? Especially once we have some channel features that are orthogonal to the commitment format, the choice of which one to use will be a choice of what trade-offs you'd rather make, so I expect nodes to have different preferences here, hence I think giving them a choice would make sense.

**Comment by t-bast** [comment]
> AFAICT, in order for this to actually work (this loose negotiation layer vs explicit channel types), things need to happen in two phases (kinda like the client+server hello in TLS before the actual encrypted session creation): first both sides agree on the channel type to be used, then the initial open_channel message is sent containing the proper parameters to kick off channel funding.

I think this is a very good idea. Negotiating the `channel_type` *before* the `open_channel` message (which contains many initial parameters that actually depend on the `channel_type` chosen) would make this more robust and give implementations more flexibility.

**Comment by Roasbeef** [comment]
> I think this is a very good idea. Negotiating the channel_type before the open_channel message (which contains many initial parameters that actually depend on the channel_type chosen) would make this more robust and give implementations more flexibility.

So this is what I've been (seemingly unsuccessfully) trying to get at: by having the initiator send multiple _options_ with the receiver _picking_ one of them, an actual _negotiation_ layer is required. 

If instead the initiator sends _just what they want to use_ (either a single enum, two enums (one for channel type and one for modifiers), or an eum and a set of bits (enum is chan type, bits are modifiers)), then this negotiation layer isn't needed at all. 

The initiator knows what the receiver deems as being possible by the feature bits they've already advertised (in `innit` and `node_announcement`). The receiver can reject any incoming funding attempt it doesn't like (as normal), ideally with an `error` message that lets it succinctly describe _why_ it didn't like the funding (bad chan type, wrong values, etc).

**Comment by ariard** [comment]
One other point to consider, you might decide to accept your channel type in function of the level
of trust you're pouring in the counterparty, which ultimaltely is a node policy decision. For e.g,
I expect we'll have far more expiremntation on the fee-bumping side, like high-feerate pre-signed
commitment transaction if you have a low-scoring of this peer reliability and brace yourself to
go onchain with high odds. Or even w.r.t fancier taproot types, one of the scripts structure might
force you to commit plainly your watchtower policy, like one tapleaf per watchtower. That's okay
if you trust the counterparty but don't do this if you expect adverserial settings.

So we might not even have something like a _static_ list of preferred channels, and we should keep
the feature flags for what they're, a signal of what the mechanisms are supported by your implementation, 
not the actual channel features authorized by your node policy on an authenticated channel opening.
You might not even communicate your channel policy as a receiver ("security by obscurity"), or at
least only to already whitelisted peers. 

For those reasons, I agree with roasbeef it's better to have the sender locking the type during
a newer initial channel type announcement phase. The fact that this type is a result of previous
interactivity between funder/fundee should be deferred to a negotiation layer, of which the 
participation is optional. We could even have a new gossip message for announcing channel policy,
for e.g "i'm open to do channel (bits XYZ) with anyone", where XYZ denominate the regular, safest
channel type supported by every implementation.

File: 02-peer-protocol.md [hunk 4]
```diff
@@ -316,14 +339,21 @@ The sender:
 avoid double-spending of the funding transaction.
   - MUST set `channel_reserve_satoshis` greater than or equal to `dust_limit_satoshis` from the `open_channel` message.
   - MUST set `dust_limit_satoshis` less than or equal to `channel_reserve_satoshis` from the `open_channel` message.
+  - if it sets `channel_type`:
+    - MUST set it to one of the `channel_types` from `open_channel`
+    - SHOULD set it to its preferred `channel_type`, or the first in `channel_types` if there is more than one equal preference.
```
Comment (pm47) [inline 02-peer-protocol.md:344]: Nit: the last part derives from the rest and seems a bit too verbose, in fact the whole line could be removed imo

```suggestion
    - SHOULD set it to its preferred `channel_type`.
```

**Review by t-bast — COMMENTED** [review]
ACK https://github.com/lightningnetwork/lightning-rfc/pull/880/commits/745b4d058cfa2c60852e933b59303d2615fbb5a8

This is implemented in https://github.com/ACINQ/eclair/pull/1867, I can test cross-implementation compatibility when it's ready in other impls, don't hesitate to ping me.

**Comment by rustyrussell** [comment]
Hmm, maybe removing the subtype was a bad idea: we actually do want an array of them for channel_upgrade.  I guess we can use a subtype there, but it's messy.

File: 02-peer-protocol.md (hunk 4 above)
Comment (rustyrussell) [reply 02-peer-protocol.md]: Yeah, we got rid of this "array of channel_types" anyway, so it's even less verbose now :)

**Comment by t-bast** [comment]
> Hmm, maybe removing the subtype was a bad idea: we actually do want an array of them for channel_upgrade. I guess we can use a subtype there, but it's messy.

But for `channel_upgrade`, this won't be in the `open_channel` or `accept_channel` message, so it doesn't matter if it's different. The way I see it `upgrade_channel` will indeed have a different tlv (e.g. `channel_types`) which is an array of `channel_type` (the same way your first proposal had an array in `open_channel` and a single value in `accept_channel`).

**Comment by rustyrussell** [comment]
> > Hmm, maybe removing the subtype was a bad idea: we actually do want an array of them for channel_upgrade. I guess we can use a subtype there, but it's messy.
> 
> But for `channel_upgrade`, this won't be in the `open_channel` or `accept_channel` message, so it doesn't matter if it's different. The way I see it `upgrade_channel` will indeed have a different tlv (e.g. `channel_types`) which is an array of `channel_type` (the same way your first proposal had an array in `open_channel` and a single value in `accept_channel`). It doesn't feel messy to me?

Well, we generate directly from the spec.  So that geneated a "channel_type", and it was easy to pass that everywhere (including replumbing the rest of our internal logic to express things as a "channel_type" rather than two bools (option_anchor_commitement and option_static_remotekey) everywhere.  Now we have a bundle of bits, it's not as clean.  However, I ended up creating a "channel_type" internally and just converting to and from the wire format, so it's not as bad as I feared.

**Review by mrkingfoxx — APPROVED** [review]

**Comment by t-bast** [comment]
I have tested cross-compatibility between eclair (https://github.com/ACINQ/eclair/pull/1867) and lnd (https://github.com/lightningnetwork/lnd/pull/5373), so this is looking good!

I had previously tested compatibility with c-lightning (https://github.com/ElementsProject/lightning/pull/4616) and found a few issues, but they seem to be mostly implementation issues and unrelated to the spec itself.

**Comment by t-bast** [comment]
Given that we have ACKs and cross-compatibility tests, should we merge this PR? @rustyrussell @Roasbeef @TheBlueMatt

**Review by t-bast — APPROVED** [review]
ACK https://github.com/lightningnetwork/lightning-rfc/pull/880/commits/745b4d058cfa2c60852e933b59303d2615fbb5a8

File: 02-peer-protocol.md [hunk 5]
```diff
@@ -217,6 +233,10 @@ The sending node:
     - MAY include `upfront_shutdown_script`.
   - if it includes `open_channel_tlvs`:
     - MUST include `upfront_shutdown_script`.
+  - if it includes `channel_type`:
+    - MUST set it to a defined type representing the type it wants.
+    - MUST use the smallest bitmap possible to represent the channel tyoe.
```
Comment (ariard) [inline 02-peer-protocol.md]: nit: s/tyoe/type/g

File: 02-peer-protocol.md [hunk 6]
```diff
@@ -252,6 +272,7 @@ are not valid secp256k1 pubkeys in compressed format.
   - the funder's amount for the initial commitment transaction is not sufficient for full [fee payment](03-transactions.md#fee-payment).
   - both `to_local` and `to_remote` amounts for the initial commitment transaction are less than or equal to `channel_reserve_satoshis` (see [BOLT 3](03-transactions.md#commitment-transaction-outputs)).
   - `funding_satoshis` is greater than or equal to 2^24 and the receiver does not support `option_support_large_channel`. 
+  - It supports `channel_type`, `channel_type` was set, and the `type` is not suitable.
```
Comment (ariard) [inline 02-peer-protocol.md:275]: " - `channel type` is set to a type containing a feature which was not negotiated" ?

File: 02-peer-protocol.md [hunk 7]
```diff
@@ -197,6 +200,19 @@ know this node will accept `funding_satoshis` greater than or equal to 2^24.
...
+Channel types are an explicit enumeration: for convenience of future
+definitions they reuse even feature bits, but they are not an
+arbitrary combination (they represent the persistent features which
+affect the channel operation).
+
+The currently defined types are:
+  - no features (no bits set)
+  - `option_static_remotekey` (bit 12)
+  - `option_anchor_outputs` and `option_static_remotekey` (bits 20 and 12)
+  - `option_anchors_zero_fee_htlc_tx` and `option_static_remotekey` (bits 22 and 12)
```
Comment (Roasbeef) [inline 02-peer-protocol.md:214]: See this comment: https://github.com/lightningnetwork/lightning-rfc/pull/824

File: 02-peer-protocol.md [hunk 8]
```diff
@@ -197,6 +200,19 @@ know this node will accept `funding_satoshis` greater than or equal to 2^24.
...
 
+#### Defined Channel Types
+
+Channel types are an explicit enumeration: for convenience of future
+definitions they reuse even feature bits, but they are not an
+arbitrary combination (they represent the persistent features which
+affect the channel operation).
+
+The currently defined types are:
+  - no features (no bits set)
```
Comment (Roasbeef) [inline 02-peer-protocol.md:211]: From my reading of the behavior here, you _must_ set a bit, as otherwise, which channel type is used? Seems that would fall back to the implicit negotiation, which is what we're trying to avoid here in the first place...

**Review by Roasbeef — APPROVED** [review]
LGTM 🎊

**Comment by Roasbeef** [comment]
@t-bast I think this can land, we should also merged in #824 (guess we just forgot to as I think many of us impl it now)?

**Comment by t-bast** [comment]
@Roasbeef I still haven't started #824, I wanted to have a good enough starting point for fee-bumping, but I'll get to it soon to unblock merging

File: 02-peer-protocol.md (hunk 8 above)
Comment (t-bast) [reply 02-peer-protocol.md:211]: Nothing set is simply the LN 1.0 channel type (before `static_remotekey`)

**Comment by Roasbeef** [comment]
@t-bast cool sgtm. 

Was rebasing our PR and realized that we had selected an interim feature bit (2020/2021) in our PR, but we don't have one here. 

This gets back to the meta conversation we were having on IRC earlier today during the spec meeting: when should new feature bits be added? I think this falls under the "new behavior that can be helpful in selecting potential channel peers for a node". If a peer I'm connected to doesn't signal this feature bit, then I have no way of knowing if they'll be able to handle the new explicit chan type negotiation.

**Comment by Roasbeef** [comment]
Related comment: https://github.com/lightningnetwork/lnd/pull/5373#issuecomment-908754143 

If we added a feature bit for something like upfront shutdown (that modified the chan negotiation messages to add new behavior), why wouldn't we add one here for something that revamps the channel type selection negotiation entirely?

File: 02-peer-protocol.md (hunk 6 above)
Comment (rustyrussell) [reply 02-peer-protocol.md:275]: You don't *need* to do additional checks: having "MUST" here (above) implies you should be testing features.

It's weird, but you should only be policing your own options, *not* theirs.  So if they offer you something you understand, you can take it.

Comment (rustyrussell) [reply 02-peer-protocol.md:275]: > Related comment: [lightningnetwork/lnd#5373 (comment)](https://github.com/lightningnetwork/lnd/pull/5373#issuecomment-908754143)
> 
> If we added a feature bit for something like upfront shutdown (that modified the chan negotiation messages to add new behavior), why wouldn't we add one here for something that revamps the channel type selection negotiation entirely?

This simply allows downgrade from "all the features we negotiated".  If we want to make this required for some new feature we add, that's cool (I think it should be for stuff like option_simplified_commitment, which you might want to support, but not prefer), but we can add that when we add the feature.

I think it's gonna be moot since everyone wants this and in 12 months it'll be compulsory anyway.... (hope)!

**Comment by rustyrussell** [comment]
Squashed into a single commit, trivial rebase on master.

Applying now as per meeting 2021-08-30.

**Comment by Roasbeef** [comment]
> I think it's gonna be moot since everyone wants this and in 12 months it'll be compulsory anyway.... (hope)!

I don't understand this new direction of "we don't actually need feature bits since everyone is going to implement it". This completely misses the purpose of feature bits which is to signal _understanding_ of new behavior and also let nodes _discover_ other nodes that also understand that behavior. 

I think we also thought that "everyone" would have implemented full anchors by now, but that's not the reality, and thankfully due to the **feature bit**, peers that understand it are able to seek out other peers that don't. 

>  If we want to make this required for some new feature we add, that's cool 

Sure and we can carve out that path _today_ by adding the optional feature bit. Once again, there's literally _zero_ cost to doing this, instead things are just more explicit and well defined. 

Without a feature bit it's essentially guess and check: you send out a new field and then attempt to see if they understood it or not. With a feature bit, you know _up front_ if they support some new behavior or not.
